boards/shared/Src/Io/Io_SharedFrequencyOnlyPwmInput.c

```diff
@@ -0,0 +1,122 @@
+#include <main.h>
+#include <math.h>
+#include "App_SharedAssert.h"
+#include "Io_SharedFrequencyOnlyPwmInput.h"
+
+#define MAX_TIMER_PERIOD_ELAPSED_COUNT 2
+
+struct FrequencyOnlyPwmInput
+{
+    float              frequency_hz;
+    TIM_HandleTypeDef *htim;
+    float              timer_frequency_hz;
+    uint32_t           rising_edge_tim_channel;
+};
+
+/**
+ * Set the frequency for the given PWM input
+ * @param pwm_input: The PWM input to set frequency for
+ * @param frequency_hz: Frequency, in Hz
+ */
+static void Io_SetFrequency(
+    struct FrequencyOnlyPwmInput *const frequency_only_pwm_input,
+    const float                         frequency_hz);
+
+static void Io_SetFrequency(
+    struct FrequencyOnlyPwmInput *const frequency_only_pwm_input,
+    const float                         frequency_hz)
+{
+    assert(frequency_hz >= 0.0f || isnanf(frequency_hz));
+
+    frequency_only_pwm_input->frequency_hz = frequency_hz;
+}
+
+struct FrequencyOnlyPwmInput *Io_SharedFrequencyOnlyPwmInput_Create(
+    TIM_HandleTypeDef *htim,
+    float              timer_frequency_hz,
+    uint32_t           rising_edge_timer_channel)
+{
+    assert(htim != NULL);
+    static struct FrequencyOnlyPwmInput
+                  freq_only_pwm_inputs[MAX_NUM_OF_FREQ_ONLY_PWM_INPUT];
+    static size_t alloc_index = 0;
+
+    assert(alloc_index < MAX_NUM_OF_FREQ_ONLY_PWM_INPUT);
+
+    struct FrequencyOnlyPwmInput *const frequency_only_pwm_input =
+        &freq_only_pwm_inputs[alloc_index++];
+
+    frequency_only_pwm_input->htim               = htim;
+    frequency_only_pwm_input->timer_frequency_hz = timer_frequency_hz;
+    frequency_only_pwm_input->rising_edge_tim_channel =
+        rising_edge_timer_channel;
+
+    HAL_TIM_IC_Start_IT(htim, rising_edge_timer_channel);
+    HAL_TIM_Base_Start_IT(htim);
+
+    return frequency_only_pwm_input;
+}
+
+void Io_SharedFrequencyOnlyPwmInput_Tick(
+    struct FrequencyOnlyPwmInput *const frequency_only_pwm_input,
+    bool *const                         first_tick,
+    uint32_t *const                     current_rising_edge,
+    uint32_t *const                     previous_rising_edge)
+{
+    if (*first_tick)
+    {
+        *current_rising_edge = HAL_TIM_ReadCapturedValue(
+            frequency_only_pwm_input->htim,
+            frequency_only_pwm_input->rising_edge_tim_channel);
+        *first_tick = false;
+    }
+    else
+    {
+        *first_tick           = true;
+        *previous_rising_edge = *current_rising_edge;
+        *current_rising_edge  = HAL_TIM_ReadCapturedValue(
+            frequency_only_pwm_input->htim,
+            frequency_only_pwm_input->rising_edge_tim_channel);
+
+        // Check if the counter value has overflowed
+        uint32_t rising_edge_delta =
```

Comment:
This is quite hard to read, can we please turn it into if-statement?

Reply:
Done 🅱️